File: src/spatial/orbit.ts
import type { Moon, Planet, StarSystem, Vector3Tuple } from '../domain/universe'

/**
 * Canonical parent-local orbit definition shared with the SpacetimeDB module.
 *
 * Coordinate convention:
 * - +Y is the reference-plane normal.
 * - The unrotated ellipse lies in X/Z.
 * - All angles are radians.
 * - `semiMajorAxis` and returned positions use the parent's local units.
 * - `phaseAtEpoch` is the eccentric-anomaly-like path parameter at simulation t=0.
 *   This intentionally gives simple deterministic motion rather than physical Kepler timing.
 */
export interface OrbitDefinition {
  semiMajorAxis: number
  eccentricity: number
  inclination: number
  longitudeOfAscendingNode: number
  argumentOfPeriapsis: number
  phaseAtEpoch: number
  angularSpeed: number
}

export interface SpatialNode {
  id: string
  parentId?: string
  /** Static parent-local offset, in the parent's local units. */
  localPosition?: Vector3Tuple
  /** Optional parent-local trajectory. Added to `localPosition` when present. */
  orbit?: OrbitDefinition
}

const TAU = Math.PI * 2

export function normalizeRadians(value: number): number {
  const normalized = value % TAU
  return normalized < 0 ? normalized + TAU : normalized
}

export function orbitPhaseAtTime(orbit: OrbitDefinition, simulationTimeSeconds: number): number {
  return normalizeRadians(orbit.phaseAtEpoch + orbit.angularSpeed * simulationTimeSeconds)
}

/** Evaluate a parent-local position using the exact same operation order as the Rust backend. */
export function orbitPositionAtPhase(
  orbit: OrbitDefinition,
  phase: number,
  renderScale = 1,
): Vector3Tuple {
  const eccentricity = Math.min(Math.max(orbit.eccentricity, 0), 0.95)
  const semiMajorAxis = Math.max(orbit.semiMajorAxis, 0) * renderScale
  const semiMinorAxis = semiMajorAxis * Math.sqrt(1 - eccentricity * eccentricity)

  // Ellipse with the parent at one focus.
  const x = semiMajorAxis * (Math.cos(phase) - eccentricity)
  const z = semiMinorAxis * Math.sin(phase)

  // 1) Rotate periapsis inside the orbital plane, around +Y.
  const cosArgument = Math.cos(orbit.argumentOfPeriapsis)
  const sinArgument = Math.sin(orbit.argumentOfPeriapsis)
  const argumentX = cosArgument * x + sinArgument * z
  const argumentZ = -sinArgument * x + cosArgument * z

  // 2) Tilt the plane around +X.
  const cosInclination = Math.cos(orbit.inclination)
  const sinInclination = Math.sin(orbit.inclination)
  const tiltedX = argumentX
  const tiltedY = -sinInclination * argumentZ
  const tiltedZ = cosInclination * argumentZ

  // 3) Rotate the line of nodes around the parent's +Y axis.
  const cosLongitude = Math.cos(orbit.longitudeOfAscendingNode)
  const sinLongitude = Math.sin(orbit.longitudeOfAscendingNode)

  return [
    cosLongitude * tiltedX + sinLongitude * tiltedZ,
    tiltedY,
    -sinLongitude * tiltedX + cosLongitude * tiltedZ,
  ]
}

export function orbitPositionAtTime(
  orbit: OrbitDefinition,
  simulationTimeSeconds: number,
  renderScale = 1,
): Vector3Tuple {
  return orbitPositionAtPhase(orbit, orbitPhaseAtTime(orbit, simulationTimeSeconds), renderScale)
}

export function sampleOrbit(
  orbit: OrbitDefinition,
  segments = 128,
  renderScale = 1,
): Vector3Tuple[] {
  const safeSegments = Math.max(8, Math.floor(segments))
  return Array.from({ length: safeSegments + 1 }, (_, index) =>
    orbitPositionAtPhase(orbit, (index / safeSegments) * TAU, renderScale),
  )
}

export function resolveSpatialNodePosition(
  nodes: ReadonlyMap<string, SpatialNode>,
  nodeId: string,
  simulationTimeSeconds: number,
): Vector3Tuple {
  const visiting = new Set<string>()

  function resolve(id: string): Vector3Tuple {
    if (visiting.has(id)) throw new Error(`Spatial node cycle detected at ${id}`)
    const node = nodes.get(id)
    if (!node) throw new Error(`Unknown spatial node: ${id}`)

    visiting.add(id)
    const staticPosition = node.localPosition ?? [0, 0, 0]
    const orbitPosition = node.orbit
      ? orbitPositionAtTime(node.orbit, simulationTimeSeconds)
      : [0, 0, 0] as Vector3Tuple
    const local: Vector3Tuple = [
      staticPosition[0] + orbitPosition[0],
      staticPosition[1] + orbitPosition[1],
      staticPosition[2] + orbitPosition[2],
    ]

    const result = node.parentId
      ? (() => {
          const parent = resolve(node.parentId!)
          return [parent[0] + local[0], parent[1] + local[1], parent[2] + local[2]] as Vector3Tuple
        })()
      : local

    visiting.delete(id)
    return result
  }

  return resolve(nodeId)
}

export function planetOrbitDefinition(planet: Planet): OrbitDefinition {
  return {
    semiMajorAxis: planet.orbitRadius,
    eccentricity: planet.orbitEccentricity ?? 0,
    inclination: planet.orbitInclination ?? 0,
    longitudeOfAscendingNode: planet.orbitLongitude ?? 0,
    argumentOfPeriapsis: planet.orbitArgument ?? 0,
    phaseAtEpoch: planet.orbitOffset,
    angularSpeed: planet.orbitSpeed,
  }
}

export function moonOrbitDefinition(moon: Moon): OrbitDefinition {
  // Keep this evaluator compatible with projects whose generated Moon type
  // predates the optional eccentricity/ascending-node columns.
  const orbitMoon = moon as Moon & {
    orbitEccentricity?: number
    orbitLongitude?: number
  }

  return {
    semiMajorAxis: moon.orbitRadius,
    eccentricity: orbitMoon.orbitEccentricity ?? 0,
    inclination: moon.orbitInclination ?? 0,
    longitudeOfAscendingNode: orbitMoon.orbitLongitude ?? 0,
    argumentOfPeriapsis: moon.orbitArgument ?? 0,
    phaseAtEpoch: moon.orbitOffset,
    angularSpeed: moon.orbitSpeed,
  }
}


/** Build the current system/planet/moon data into the generic spatial-node graph. */
export function buildSystemSpatialNodes(system: StarSystem): Map<string, SpatialNode> {
  const nodes = new Map<string, SpatialNode>()
  nodes.set(system.id, {
    id: system.id,
    localPosition: [0, 0, 0],
  })

  for (const planet of system.planets) {
    nodes.set(planet.id, {
      id: planet.id,
      parentId: system.id,
      orbit: planetOrbitDefinition(planet),
    })

    for (const moon of planet.moons ?? []) {
      nodes.set(moon.id, {
        id: moon.id,
        parentId: planet.id,
        orbit: moonOrbitDefinition(moon),
      })
    }
  }

  return nodes
}

File: spacetimedb/src/spatial/orbit.rs
use crate::tables::data_tables::{Moon, Planet, Vec3f};

#[derive(Clone, Copy, Debug)]
pub struct OrbitDefinition {
    /// Semi-major axis in the parent node's local units.
    pub semi_major_axis: f64,
    pub eccentricity: f64,
    /// All orbital angles are radians.
    pub inclination: f64,
    pub longitude_of_ascending_node: f64,
    pub argument_of_periapsis: f64,
    /// Eccentric-anomaly-like path parameter at simulation t=0.
    pub phase_at_epoch: f64,
    /// Radians per simulation second.
    pub angular_speed: f64,
}

const TAU: f64 = std::f64::consts::TAU;


pub fn planet_orbit_definition(planet: &Planet) -> OrbitDefinition {
    OrbitDefinition {
        semi_major_axis: f64::from(planet.orbit_radius),
        eccentricity: f64::from(planet.orbit_eccentricity),
        inclination: f64::from(planet.orbit_inclination),
        longitude_of_ascending_node: f64::from(planet.orbit_longitude),
        argument_of_periapsis: f64::from(planet.orbit_argument),
        phase_at_epoch: f64::from(planet.orbit_offset),
        angular_speed: f64::from(planet.orbit_speed),
    }
}

pub fn moon_orbit_definition(moon: &Moon) -> OrbitDefinition {
    OrbitDefinition {
        semi_major_axis: f64::from(moon.orbit_radius),
        eccentricity: f64::from(moon.orbit_eccentricity),
        inclination: f64::from(moon.orbit_inclination),
        longitude_of_ascending_node: f64::from(moon.orbit_longitude),
        argument_of_periapsis: f64::from(moon.orbit_argument),
        phase_at_epoch: f64::from(moon.orbit_offset),
        angular_speed: f64::from(moon.orbit_speed),
    }
}

pub fn normalize_radians(value: f64) -> f64 {
    value.rem_euclid(TAU)
}

pub fn orbit_phase_at_time(orbit: OrbitDefinition, simulation_time_seconds: f64) -> f64 {
    normalize_radians(orbit.phase_at_epoch + orbit.angular_speed * simulation_time_seconds)
}

/// Parent-local orbit evaluation. Keep this operation order identical to
/// `src/spatial/orbit.ts` so reducers and renderers agree exactly.
pub fn orbit_position_at_phase(orbit: OrbitDefinition, phase: f64) -> Vec3f {
    let eccentricity = orbit.eccentricity.clamp(0.0, 0.95);
    let semi_major_axis = orbit.semi_major_axis.max(0.0);
    let semi_minor_axis = semi_major_axis * (1.0 - eccentricity * eccentricity).sqrt();

    let x = semi_major_axis * (phase.cos() - eccentricity);
    let z = semi_minor_axis * phase.sin();

    let cos_argument = orbit.argument_of_periapsis.cos();
    let sin_argument = orbit.argument_of_periapsis.sin();
    let argument_x = cos_argument * x + sin_argument * z;
    let argument_z = -sin_argument * x + cos_argument * z;

    let cos_inclination = orbit.inclination.cos();
    let sin_inclination = orbit.inclination.sin();
    let tilted_x = argument_x;
    let tilted_y = -sin_inclination * argument_z;
    let tilted_z = cos_inclination * argument_z;

    let cos_longitude = orbit.longitude_of_ascending_node.cos();
    let sin_longitude = orbit.longitude_of_ascending_node.sin();

    Vec3f {
        x: (cos_longitude * tilted_x + sin_longitude * tilted_z) as f32,
        y: tilted_y as f32,
        z: (-sin_longitude * tilted_x + cos_longitude * tilted_z) as f32,
    }
}

pub fn orbit_position_at_time(orbit: OrbitDefinition, simulation_time_seconds: f64) -> Vec3f {
    orbit_position_at_phase(orbit, orbit_phase_at_time(orbit, simulation_time_seconds))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn assert_close(actual: f32, expected: f32) {
        assert!((actual - expected).abs() < 0.000_1, "{actual} != {expected}");
    }

    #[test]
    fn circular_unrotated_orbit_starts_on_positive_x() {
        let orbit = OrbitDefinition {
            semi_major_axis: 10.0,
            eccentricity: 0.0,
            inclination: 0.0,
            longitude_of_ascending_node: 0.0,
            argument_of_periapsis: 0.0,
            phase_at_epoch: 0.0,
            angular_speed: 1.0,
        };

        let position = orbit_position_at_time(orbit, 0.0);
        assert_close(position.x, 10.0);
        assert_close(position.y, 0.0);
        assert_close(position.z, 0.0);
    }

    #[test]
    fn inclination_produces_vertical_motion() {
        let orbit = OrbitDefinition {
            semi_major_axis: 10.0,
            eccentricity: 0.2,
            inclination: 0.3,
            longitude_of_ascending_node: 0.7,
            argument_of_periapsis: 0.4,
            phase_at_epoch: std::f64::consts::FRAC_PI_2,
            angular_speed: 0.0,
        };

        let position = orbit_position_at_time(orbit, 0.0);
        assert!(position.y.abs() > 0.1);
    }
}
